torch/csrc/jit/codegen/cuda/python_frontend/fusion_record.h

```diff
@@ -0,0 +1,256 @@
+#pragma once
+#include <c10/util/complex.h>
+#include <torch/csrc/jit/codegen/cuda/arith.h>
+#include <torch/csrc/jit/codegen/cuda/ops/normalization.h>
+#include <torch/csrc/jit/codegen/cuda/python_frontend/fusion_definition.h>
+
+namespace nvfuser {
+struct RecordFunctor {
+  RecordFunctor(std::vector<size_t> _args, std::vector<size_t> _outputs)
+      : args(std::move(_args)), outputs(std::move(_outputs)) {}
+  virtual void operator()(FusionDefinition& fd) = 0;
+
+  std::vector<size_t> args;
+  std::vector<size_t> outputs;
+};
+
+struct InputTensorRecord : RecordFunctor {
+  InputTensorRecord(
+      std::vector<size_t> _outputs,
+      std::vector<int64_t> _symbolic_sizes,
+      std::vector<bool> _contiguous_info,
+      NvfDataType _dtype)
+      : RecordFunctor({}, std::move(_outputs)),
+        symbolic_sizes(std::move(_symbolic_sizes)),
+        contiguous_info(std::move(_contiguous_info)),
+        dtype(_dtype) {}
+  void operator()(FusionDefinition& fd) final {
+    auto tv = TensorViewBuilder()
+                  .ndims(symbolic_sizes.size())
+                  .contiguity(contiguous_info)
+                  .shape(symbolic_sizes)
+                  .dtype(dtype)
+                  .build();
+
+    fd.fusion_state.at(outputs.at(0)) = tv;
+    fd.addInput(tv);
+  }
+
+  std::vector<int64_t> symbolic_sizes;
+  std::vector<bool> contiguous_info;
+  NvfDataType dtype;
+};
+
+struct ScalarRecord : RecordFunctor {
+  ScalarRecord(std::vector<size_t> _outputs, NvfDataType dtype)
+      : RecordFunctor({}, std::move(_outputs)), dtype_(dtype) {}
+
+  void operator()(FusionDefinition& fd) final {
+    NvfVal* output = nullptr;
+    if (dtype_ == NvfDataType::Double) {
+      output = IrBuilder::create<torch::jit::fuser::cuda::Double>();
+    } else if (dtype_ == NvfDataType::ComplexDouble) {
+      output = IrBuilder::create<torch::jit::fuser::cuda::ComplexDouble>();
+    } else if (dtype_ == NvfDataType::Bool) {
+      output = IrBuilder::create<torch::jit::fuser::cuda::Bool>();
+    } else if (dtype_ == NvfDataType::Int) {
+      output = IrBuilder::create<torch::jit::fuser::cuda::Int>();
+    } else {
+      TORCH_CHECK(false, "Dtype is not supported:", dtype_);
+    }
+    fd.addInput(output);
+    fd.fusion_state.at(outputs.at(0)) = output;
+  }
+
+ private:
+  NvfDataType dtype_;
+};
+
+template <typename ExprType, typename ValueType>
+struct ConstantRecord : RecordFunctor {
+  ConstantRecord(std::vector<size_t> _outputs, ValueType val)
+      : RecordFunctor({}, std::move(_outputs)), value_(val) {}
+
+  void operator()(FusionDefinition& fd) final {
+    NvfVal* output = IrBuilder::create<ExprType>(value_);
+    fd.fusion_state.at(outputs.at(0)) = output;
+  }
+
+ private:
+  ValueType value_;
+};
+
+template <class OutputType>
+struct OutputRecord : RecordFunctor {
```

Comment:
ditto. Should we consider aggregating all outputs to a single OutputRecord node?

Reply:
See comment above.